sw/device/lib/crypto/drivers/kmac.c

```diff
@@ -656,3 +665,107 @@ kmac_error_t kmac_process_msg_blocks(kmac_operation_t operation,
 
   return kKmacOk;
 }
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_sha3_224(const uint8_t *data, size_t data_len,
+                           uint8_t *digest) {
+  if (data == NULL && data_len > 0) {
+    return kKmacArgsError;
+  } else if (digest == NULL) {
+    return kKmacArgsError;
+  }
+
+  kmac_error_t err;
+  err = kmac_init(kKmacOperationSHA3, kKmacSecurityStrength224,
+                  /*func_name=*/NULL, /*func_name_len=*/0,
+                  /*cust_str=*/NULL, /*cust_str_len=*/0);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  err = kmac_process_msg_blocks(kKmacOperationSHA3, data, data_len, digest,
+                                /*digest_len=*/224 / 8);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  return kKmacOk;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_sha3_256(const uint8_t *data, size_t data_len,
+                           uint8_t *digest) {
+  if (data == NULL && data_len > 0) {
+    return kKmacArgsError;
+  } else if (digest == NULL) {
+    return kKmacArgsError;
+  }
+
+  kmac_error_t err;
+  err = kmac_init(kKmacOperationSHA3, kKmacSecurityStrength256,
+                  /*func_name=*/NULL, /*func_name_len=*/0,
+                  /*cust_str=*/NULL, /*cust_str_len=*/0);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  err = kmac_process_msg_blocks(kKmacOperationSHA3, data, data_len, digest,
+                                /*digest_len=*/256 / 8);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  return kKmacOk;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_sha3_384(const uint8_t *data, size_t data_len,
+                           uint8_t *digest) {
+  if (data == NULL && data_len > 0) {
+    return kKmacArgsError;
+  } else if (digest == NULL) {
+    return kKmacArgsError;
+  }
+
+  kmac_error_t err;
+  err = kmac_init(kKmacOperationSHA3, kKmacSecurityStrength384,
+                  /*func_name=*/NULL, /*func_name_len=*/0,
+                  /*cust_str=*/NULL, /*cust_str_len=*/0);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  err = kmac_process_msg_blocks(kKmacOperationSHA3, data, data_len, digest,
+                                /*digest_len=*/384 / 8);
+  if (err != kKmacOk) {
+    return err;
+  }
+
+  return kKmacOk;
+}
+
+OT_WARN_UNUSED_RESULT
+kmac_error_t kmac_sha3_512(const uint8_t *data, size_t data_len,
+                           uint8_t *digest) {
+  if (data == NULL && data_len > 0) {
+    return kKmacArgsError;
+  } else if (digest == NULL) {
+    return kKmacArgsError;
+  }
```

Comment:
For this and other functions, should parameter checks be implemented in `hash.c`? This seems to be a private function (not exposed at the API level), and we'll have to be careful about code size. It may be a good idea to keep these checks for hardening purposes, but then it would be good to consider collapsing all the sha3 functions to optimize for code size.

Reply:
So far, I duplicated many checks for hardening and easier debugging. If we do not need this type of hardening, I think it makes sense to move all of them to the caller functions in `hash.c`. I was not aware that code size is an issue.

By collapsing SHA functions for smaller code size, do you mean defining single `kmac_sha3(...)` which takes which SHA-3 algorithm is used? Or just making the body of  these split `kmac_sha3_<XYZ>` functions smaller by removing the error checks in them?

Reply:
Speaking of public/private functions, would you like to define a convention where "private" functions are declared in a separate header file, e.g. `kmac_private.h`?

Reply:
I got a bit confused about what we refer to as code size. I initially though it was the compiled binary image size, but now I think we are talking about keeping the code organized and not hoarding all KMAC-related functions into the same file (that is `kmac.c`).

I guess one way to split `kmac.c` is to move functions like `kmac_sha3_224` (which are added in this PR) somewhere else and treat them as if they are public (with no error checks). Is this also what you meant @alphan? Just to clarify, there would be two files `kmac_private.h`, `kmac_private.c` and the latter would contain the functions, right?

Reply:
> I got a bit confused about what we refer to as code size. I initially though it was the compiled binary image size, but now I think we are talking about keeping the code organized and not hoarding all KMAC-related functions into the same file (that is `kmac.c`).

You're right, we tend to use code-/binary-size interchangeably to refer to the binary size. My comment about public and private functions was inspired by @moidx observation. 

> I guess one way to split `kmac.c` is to move functions like `kmac_sha3_224` (which are added in this PR) somewhere else and treat them as if they are public (with no error checks). Is this also what you meant @alphan? Just to clarify, there would be two files `kmac_private.h`, `kmac_private.c` and the latter would contain the functions, right?

I'm not super familiar with the API boundaries but to keep things simple, you might want to introduce `kmac_private.h` for things that you want to easily test but not expose in the public APIs while keeping all implementations in a single file `kmac.c`.

Reply:
I don't think there's a need for another layer here -- `kmac.c` is already the private layer, and only the cryptolib can call it. The only functions exposed in the crypto library are the ones under `sw/device/lib/crypto/include` (in this case  `sw/device/lib/crypto/include/hash.h`). Nothing else is exposed.

Reply:
Just to wrap up this thread, here is what I will be doing:

- Remove NULL pointer and arguments checks from "private" `kmac.c` code. I will be checking those arguments in "public" `hash.c` instead.
- Keep `kmac.c` as it is, because it contains only "private" functions.
- All tests will eventually start using "public" functions in `hash.c`. For this PR, this will be an exception: `test_kmac.c` is calling private functions from `kmac.c`. I will remove all of those when all SHA-3/SHAKE/cSHAKE/KMAC are connected to public cryptolib API. In the end, all tests should go through public cryptolib API.

Let me know if any of those does not make sense. I will repeat the same pattern for the follow-up PRs for SHAKE/cSHAKE/KMAC.

Reply:
SGTM. It's ok to have tests for functions that we consider "private" so `test_kmac.c` can stay.